### PR TITLE
[LoRA] Prepare Punica metadata on CPU

### PR DESCRIPTION
**Pull request by cr-zhao**
## Summary

- prepare Punica LoRA kernel metadata from the CPU-side mapping
- use Python bucketing for small mappings and vectorized PyTorch operations for large mappings
- copy the prepared metadata to the existing device buffers with non-blocking H2D copies
- reuse prepared metadata for identical decode token and prompt mappings
- remove the `gpu_sync_allowed()` exception from `PunicaWrapperGPU.update_metadata()`
- cover mixed-LoRA, active-LoRA specialization, and no-LoRA metadata

## Why

This follows up on the TODO left by #43107. `PunicaWrapperGPU.update_metadata()` previously passed device mappings to `LoRAKernelMeta.prepare_tensors()`. Its no-LoRA check then consumed a CUDA scalar from Python and copied it into a CPU tensor, forcing a GPU-to-CPU synchronization on the main stream.

The mapping originates as Python/CPU data. This change retains the CPU LoRA indices during mapping conversion and prepares the sort, grouping, count, and start-location metadata on CPU. Small mappings use Python bucketing, while large mappings use vectorized PyTorch operations. The prepared metadata is copied asynchronously into the stable device buffers, and identical decode mappings share one preparation pass. This avoids reading metadata back from the GPU and preserves the buffer addresses used by CUDA graphs.

I searched open PRs for `43107`, `PunicaWrapperGPU update_metadata`, and `LoRA metadata GPU CPU sync`. I did not find another PR addressing this TODO. The related open LoRA slab PR optimizes adapter weight transfers rather than per-step Punica kernel metadata.

## Validation

- `.venv/bin/python -m compileall -q vllm/lora/punica_wrapper vllm/lora/ops/triton_ops/lora_kernel_metadata.py tests/lora/test_punica_ops.py`
- `git diff --check`
- `pytest tests/lora/test_punica_ops.py -k early_exit -q` could not start in the current environment because the existing virtual environment is missing `tblib` (and subsequently `numpy`); no pytest result is claimed.

Model evaluation is not applicable because this only changes how equivalent LoRA kernel metadata is prepared and transferred.

OpenAI Codex was used to inspect the synchronization path, implement the change, and draft tests and this description. I reviewed each line.

### CODE REVIEW

File: vllm/lora/punica_wrapper/utils.py [hunk 1]
```diff
@@ -120,9 +130,21 @@ def convert_mapping(
...
+        dtype=torch.long,
+        device="cpu",
+        pin_memory=PIN_MEMORY,
+    )
+    prompt_mapping_cpu = torch.tensor(
+        prompt_mapping,
+        dtype=torch.long,
+        device="cpu",
+        pin_memory=PIN_MEMORY,
+    )
```
Comment (njhill) [inline vllm/lora/punica_wrapper/utils.py]: No need for these, can pass to `async_tensor_h2d` directly

**Review by njhill — COMMENTED** [review]
Thanks @cr-zhao

File: vllm/lora/punica_wrapper/utils.py (hunk 1 above)
Comment (cr-zhao) [reply vllm/lora/punica_wrapper/utils.py]: Thanks for the review! I have updated the related logic accordingly.

**Comment by jeejeelee** [comment]
Can you provide the perf comparison for this?

**Comment by cr-zhao** [comment]
Yes, @jeejeelee. I ran a same-instance microbenchmark of `PunicaWrapperGPU.update_metadata()` on the PR base and the current head.

### Environment

- GPU: NVIDIA GeForce RTX 4090
- PyTorch: 2.13.0+cu132
- CUDA: 13.2
- Base: `9ff7041b55`
- PR: `cd0086871a`
- Warmup: 100 iterations
- Measured: 1,000 iterations per case
### Results

The table reports mean completion latency after CUDA synchronization, so the asynchronous H2D work is included. Lower latency is better.

| Mode | Tokens | Prompts | Active LoRAs | Base (µs) | PR (µs) | Improvement |
|---|---:|---:|---:|---:|---:|---:|
| decode | 1 | 1 | 1 | 844.16 | 494.77 | 41.4% |
| decode | 16 | 16 | 1 | 869.99 | 504.35 | 42.0% |
| decode | 16 | 16 | 4 | 868.98 | 506.72 | 41.7% |
| decode | 16 | 16 | 8 | 869.84 | 508.55 | 41.5% |
| decode | 128 | 128 | 1 | 921.36 | 578.09 | 37.3% |
| decode | 128 | 128 | 4 | 926.06 | 579.09 | 37.5% |
| decode | 128 | 128 | 8 | 924.24 | 578.87 | 37.4% |
| decode | 512 | 512 | 1 | 1084.27 | 817.16 | 24.6% |
| decode | 512 | 512 | 4 | 1085.48 | 820.64 | 24.4% |
| decode | 512 | 512 | 8 | 1086.30 | 819.80 | 24.5% |
| decode | 1024 | 1024 | 1 | 1313.15 | 1053.35 | 19.8% |
| decode | 1024 | 1024 | 4 | 1311.35 | 1057.99 | 19.3% |
| decode | 1024 | 1024 | 8 | 1312.92 | 1061.43 | 19.2% |
| decode | 2048 | 2048 | 1 | 1750.80 | 1575.14 | 10.0% |
| decode | 2048 | 2048 | 4 | 1753.66 | 1575.97 | 10.1% |
| decode | 2048 | 2048 | 8 | 1750.00 | 1597.08 | 8.7% |
| prefill | 1 | 1 | 1 | 843.97 | 506.25 | 40.0% |
| prefill | 16 | 1 | 1 | 859.14 | 555.55 | 35.3% |
| prefill | 128 | 1 | 1 | 897.32 | 611.03 | 31.9% |
| prefill | 512 | 4 | 1 | 1029.93 | 818.96 | 20.5% |
| prefill | 512 | 4 | 4 | 1029.52 | 824.47 | 19.9% |
| prefill | 1024 | 8 | 1 | 1198.46 | 1025.04 | 14.5% |
| prefill | 1024 | 8 | 4 | 1199.39 | 1029.62 | 14.2% |
| prefill | 1024 | 8 | 8 | 1199.20 | 1033.22 | 13.8% |
| prefill | 2048 | 16 | 1 | 1528.51 | 1452.32 | 5.0% |
| prefill | 2048 | 16 | 4 | 1529.08 | 1464.14 | 4.2% |
| prefill | 2048 | 16 | 8 | 1526.85 | 1468.56 | 3.8% |

Mean completion latency improved in every tested configuration, ranging from 3.8% to 42.0%.

Mean and median completion latency improved in every tested configuration.
Two 2048-token, 8-active-LoRA cases showed higher P99 latency in this run:

- decode: 1791.82 µs → 1946.73 µs
- prefill: 1567.16 µs → 1634.21 µs

Additional repetitions would be useful to determine whether these tail-latency differences are measurement noise.

This is a focused metadata-path microbenchmark rather than an end-to-end model throughput benchmark.

The benchmark script and raw JSON results are available if useful:
[pr-final.json](https://github.com/user-attachments/files/31441007/pr-final.json)
[baseline-final.json](https://github.com/user-attachments/files/31441008/baseline-final.json)
[benchmark_punica_metadata.py](https://github.com/user-attachments/files/31441013/benchmark_punica_metadata.py)

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 2]
```diff
@@ -106,6 +110,107 @@ def _reset(self):
...
+                    sorted_lora_ids, return_counts=True
+                )
+                sorted_indices_cpu = sorted_indices_cpu.to(dtype=torch.int32)
+                counts_cpu = counts_cpu.to(dtype=torch.int32)
+                start_locs_cpu = torch.cumsum(counts_cpu, dim=0, dtype=torch.int32)
+                if PIN_MEMORY:
+                    sorted_indices_cpu = sorted_indices_cpu.pin_memory()
+                    lora_ids_cpu = lora_ids_cpu.pin_memory()
+                    counts_cpu = counts_cpu.pin_memory()
+                    start_locs_cpu = start_locs_cpu.pin_memory()
```
Comment (njhill) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py]: I think some copies can be avoided here by preallocating pinned tensors with the target dtypes and/or using `out` args, etc.

We shouldn't care about redundant work in the `PIN_MEMORY == False` case

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 3]
```diff
@@ -106,6 +110,107 @@ def _reset(self):
...
+        copy_to: "LoRAKernelMeta | None" = None,
+    ) -> None:
+        """Prepare metadata on the CPU and asynchronously copy it to device."""
+        no_lora = all(lora_id == -1 for lora_id in token_lora_mapping)
+        num_tokens = len(token_lora_mapping)
+        if not no_lora:
+            mapping_cpu = torch.tensor(
+                token_lora_mapping,
+                dtype=torch.int32,
+                pin_memory=PIN_MEMORY,
```
Comment (njhill) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py:126]: formatting nit:
```suggestion
                token_lora_mapping, dtype=torch.int32, pin_memory=PIN_MEMORY
```
(and similar further below)

**Comment by cr-zhao** [comment]
Since the latest changes modify logic that may affect performance, I ran an additional targeted benchmark against the previous PR commit.

### Environment

- GPU: NVIDIA GeForce RTX 4090
- PyTorch: 2.13.0+cu132
- CUDA: 13.2
- Before: `cd0086871a`
- After: `788910b306`
- Warmup: 200 iterations
- Measurement: 3,000 iterations per workload
- Workloads: decode with 1,024/2,048 tokens and 1/4/8 active LoRAs

### Results

Positive percentages indicate lower latency in the latest commit.

| Tokens | Active LoRAs | Metric | Before mean (µs) | After mean (µs) | Mean improvement | P50 improvement | P99 improvement |
|---:|---:|---|---:|---:|---:|---:|---:|
| 1,024 | 1 | dispatch | 1,031.62 | 1,011.58 | +1.94% | +4.37% | -6.57% |
| 1,024 | 1 | completion | 1,048.53 | 1,000.68 | +4.56% | +4.54% | +1.68% |
| 1,024 | 4 | dispatch | 1,037.98 | 1,040.28 | -0.22% | +0.82% | -5.89% |
| 1,024 | 4 | completion | 1,052.27 | 1,028.09 | +2.30% | +4.27% | -1.87% |
| 1,024 | 8 | dispatch | 1,038.46 | 1,004.09 | +3.31% | +4.09% | +0.50% |
| 1,024 | 8 | completion | 1,055.25 | 1,023.80 | +2.98% | +3.90% | -10.00% |
| 2,048 | 1 | dispatch | 1,575.12 | 1,509.02 | +4.20% | +4.58% | +6.09% |
| 2,048 | 1 | completion | 1,550.48 | 1,524.66 | +1.67% | +4.07% | -8.28% |
| 2,048 | 4 | dispatch | 1,542.65 | 1,499.13 | +2.82% | +4.28% | -9.80% |
| 2,048 | 4 | completion | 1,556.67 | 1,500.62 | +3.60% | +4.23% | -17.71% |
| 2,048 | 8 | dispatch | 1,529.38 | 1,496.34 | +2.16% | +3.74% | -12.02% |
| 2,048 | 8 | completion | 1,544.79 | 1,486.08 | +3.80% | +3.94% | -1.00% |

Across these targeted workloads, the latest commit reduced mean latency by 2.76% and P50 latency by 3.90% on average. Mean latency improved in 11 of the 12 cases.

P99 was noisier and regressed in several cases despite consistent mean and P50 improvements, suggesting sensitivity to occasional system-level outliers on the shared benchmark instance.

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py (hunk 2 above)
Comment (cr-zhao) [reply vllm/lora/ops/triton_ops/lora_kernel_metadata.py]: Thanks for the suggestion! I updated this to preallocate the destination tensors with the target dtypes using `torch.empty_like(..., pin_memory=PIN_MEMORY)` and write the results directly into them, also used the `out` argument for `torch.cumsum`. Passing `PIN_MEMORY` during allocation and removes the explicit `if PIN_MEMORY` branch.

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py (hunk 3 above)
Comment (cr-zhao) [reply vllm/lora/ops/triton_ops/lora_kernel_metadata.py:126]: updated the similar formatting cases

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 4]
```diff
@@ -106,6 +110,104 @@ def _reset(self):
...
+            if num_tokens >= _VECTORIZED_METADATA_MIN_TOKENS:
+                sorted_lora_ids, sorted_indices = torch.sort(mapping_cpu, stable=True)
+                lora_ids, counts = torch.unique_consecutive(
+                    sorted_lora_ids, return_counts=True
+                )
+
+                sorted_indices_cpu = torch.empty_like(
+                    sorted_indices, dtype=torch.int32, pin_memory=PIN_MEMORY
+                )
+                sorted_indices_cpu.copy_(sorted_indices)
```
Comment (njhill) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py]: Could you use the `out` arg ot `torch.sort` here to avoid this copy?

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 5]
```diff
@@ -106,6 +110,104 @@ def _reset(self):
...
+        token_lora_mapping: list[int],
+        copy_to: "LoRAKernelMeta | None" = None,
+    ) -> None:
+        """Prepare metadata on the CPU and asynchronously copy it to device."""
+        no_lora = all(lora_id == -1 for lora_id in token_lora_mapping)
+        num_tokens = len(token_lora_mapping)
+        if not no_lora:
+            mapping_cpu = torch.tensor(
+                token_lora_mapping, dtype=torch.int32, pin_memory=PIN_MEMORY
+            )
```
Comment (njhill) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py:127]:
```suggestion
            mapping_cpu = torch.tensor(
                token_lora_mapping, dtype=torch.int32, device="cpu", pin_memory=PIN_MEMORY
            )
```

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 6]
```diff
@@ -106,6 +110,104 @@ def _reset(self):
...
+                )
+
+                sorted_indices_cpu = torch.empty_like(
+                    sorted_indices, dtype=torch.int32, pin_memory=PIN_MEMORY
+                )
+                sorted_indices_cpu.copy_(sorted_indices)
+                lora_ids_cpu = torch.empty_like(
+                    lora_ids, dtype=torch.int32, pin_memory=PIN_MEMORY
+                )
+                lora_ids_cpu.copy_(lora_ids)
```
Comment (njhill) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py]: nit:
```suggestion
                lora_ids_cpu = torch.empty_like(
                    lora_ids, dtype=torch.int32, pin_memory=PIN_MEMORY
                ).copy_(lora_ids)
```

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py [hunk 7]
```diff
@@ -106,6 +110,104 @@ def _reset(self):
...
+                counts_cpu = torch.empty_like(
+                    counts, dtype=torch.int32, pin_memory=PIN_MEMORY
+                )
+                counts_cpu.copy_(counts)
```
Comment (njhill) [inline vllm/lora/ops/triton_ops/lora_kernel_metadata.py]: nit:
```suggestion
                counts_cpu = torch.empty_like(
                    counts, dtype=torch.int32, pin_memory=PIN_MEMORY
                ).copy_(counts)
```

File: vllm/lora/ops/triton_ops/lora_kernel_metadata.py (hunk 4 above)
Comment (cr-zhao) [reply vllm/lora/ops/triton_ops/lora_kernel_metadata.py]: Thanks, now use `out` to write directly into a preallocated pinned int64 tensor.

**Comment by cr-zhao** [comment]
pre-commit checks are passing.

**Comment by cr-zhao** [comment]
gentle ping @njhill @jeejeelee

**Comment by linitra24** [comment]
Hi, I ran an end-to-end comparison of the PR against the main branch on a single H20 using Qwen3-30B-A3B-FP8, 8 LoRA adapters, 4096-token inputs, 100 requests, and a request rate of 10 :

Output | Output throughput, Base → PR | Mean TTFT, Base → PR | Mean TPOT, Base → PR
-- | -- | -- | --
32 | 144.76 → 145.82 tok/s (+0.73%) | 5560.3 → 5480.7 ms (−1.43%) | 294.33 → 292.19 ms (−0.73%)
64 | 273.40 → 275.13 tok/s (+0.63%) | 5566.0 → 5492.3 ms (−1.32%) | 186.12 → 185.18 ms (−0.50%)
128 | 489.96 → 493.84 tok/s (+0.79%) | 5543.0 → 5490.4 ms (−0.95%) | 113.79 → 112.89 ms (−0.79%)
256 | 810.99 → 812.14 tok/s (+0.14%) | 5534.8 → 5496.7 ms (−0.69%) | 78.07 → 78.03 ms (-0.05%)
512 | 1194.73 → 1198.54 tok/s (+0.32%) | 5536.9 → 5502.5 ms (−0.62%) | 61.02 → 60.87 ms (−0.24%)

The retested results are consistent across output lengths. The PR is neutral to slightly positive end to end, improving output throughput by 0.14–0.79% and mean TTFT by 0.62–1.43%. Mean TPOT is also effectively unchanged or slightly improved. Since the differences are below 1.5%, they should be interpreted as modest improvements within normal benchmark variability.

**Comment by njhill** [comment]
@cr-zhao this LGTM but also waiting for approval from @jeejeelee

**Comment by cr-zhao** [comment]
@linitra24 Thanks for the additional E2E testing and detailed results! Really appreciate it.

**Comment by jeejeelee** [comment]
TBH, this improvement is pretty marginal — not sure it's worth introducing.
